java-tools/OsmAndServer/src/main/java/net/osmand/server/api/services/UserdataService.java

```diff
@@ -1195,4 +1193,125 @@ public void updateDeviceLangInfo(PremiumUserDevicesRepository.PremiumUserDevice
             devicesRepository.saveAndFlush(dev);
         }
     }
+
+	@Transactional
+	public String generateSharedUrl(PremiumUserFilesRepository.UserFile userFile) {
+		String fileName = userFile.name.replaceAll("\\s+", "_").toLowerCase();
+		String uniqueToken = UUID.randomUUID().toString();
+		String url = fileName + "_" + uniqueToken;
+		userFile.sharedUrl = url;
+		filesRepository.saveAndFlush(userFile);
+
+		return url;
+	}
```

Comment:
Using fileName in the URL is a dangerous practice that could lead to charset/encoding and security issues. Please avoid it. It is better to use file-id, possibly mixed with some short hash. The final URL should be short, sweet, and easily regenerated.